lib/datadog/di/code_tracker.rb

```diff
@@ -0,0 +1,152 @@
+        # For now, until there is a use case for recreating the trace point,
+        # do nothing if the code tracker has already started.
+        return if active?
+
+        compiled_trace_point = TracePoint.trace(:script_compiled) do |tp|
+          # Useful attributes of the trace point object here:
+          # .instruction_sequence
+          # .method_id
+          # .path (refers to the code location that called the require/eval/etc.,
+          #   not where the loaded code is; use .path on the instruction sequence
+          #   to obtain the location of the compiled code)
+          # .eval_script
+          #
+          # For now just map the path to the instruction sequence.
+          path = tp.instruction_sequence.path
+          registry_lock.synchronize do
+            registry[path] = tp.instruction_sequence
+          end
+        end
+
+        trace_point_lock.synchronize do
+          # Since trace point creation itself is not under a lock, see if
+          # another thread created the trace point, in which case we can
+          # disable our trace point and do nothing.
+          if @compiled_trace_point
+            # Disable the local variable, leave instance variable as it is.
+            compiled_trace_point.disable
+            return
+          end
+
+          @compiled_trace_point = compiled_trace_point
+        end
+      end
+
+      # Returns whether this code tracker has been activated and is
+      # tracking.
+      def active?
+        trace_point_lock.synchronize do
+          !!@compiled_trace_point
+        end
+      end
+
+      # Returns an array of RubVM::InstructionSequence (i.e. the compiled code)
+      # for the provided path.
+      #
+      # The argument can be a full path to a Ruby source code file or a
+      # suffix (basename + one or more directories preceding the basename).
+      # The idea with suffix matches is that file paths are likely to
+      # be different between development and production environments and
+      # the source control system uses relative paths and doesn't have
+      # absolute paths at all.
+      #
+      # Suffix matches are not guaranteed to be correct, meaning there may
+      # be multiple files with the same basename and they may all match a
+      # given suffix. In such cases, this method will return all matching
+      # paths (and all of these paths will be attempted to be instrumented
+      # by upstream code).
+      #
+      # If the suffix matches one of the paths completely (which requires it
+      # to be an absolute path), only the exactly matching path is returned.
+      # Otherwise all known paths that end in the suffix are returned.
+      # If no paths match, an empty array is returned.
+      def iseqs_for_path(suffix)
+        registry_lock.synchronize do
+          exact = registry[suffix]
+          if exact
+            return [exact]
+          end
+          inexact = []
+          registry.each do |path, iseq|
```

Comment:
May I suggest to make a guard-clause more visual or there are some constraints?

```suggestion
          exact = registry[suffix]
          return [exact] if exact

          inexact = []
          registry.each do |path, iseq|
```